curs/curs-18-19-linking/README.md

```diff
@@ -0,0 +1,1297 @@
+Simbolul `increment` se găsește la adresa `0x08048100` adică la începutul secțiunii `.text`.
+Simbolul `num_items` se găsește la adresa `0x0804a000` adică la începutul secțiunii `.data`.
+Simbolul `main` se găsește la adresa `0x08048113` adică la offsetul `0x13` în secțiunea `.text`.
+
+### Relocarea simbolurilor
+
+În dezasamblarea fișierului obiect `one.o`, respectiv a executabilului `one`, observăm că instrucțiunile care folosesc variabila `num_items` sunt:
+```
+; one.o
+   [...]
+   3:	a1 00 00 00 00       	mov    eax,ds:0x0
+   [...]
+   b:	a3 00 00 00 00       	mov    ds:0x0,eax
+   [...]
+  16:	c7 05 00 00 00 00 05 	mov    DWORD PTR ds:0x0,0x5
+
+; one
+ [...]
+ 8048103:	a1 00 a0 04 08       	mov    eax,ds:0x804a000
+ [...]
+ 804810b:	a3 00 a0 04 08       	mov    ds:0x804a000,eax
+ [...]
+ 8048116:	c7 05 00 a0 04 08 05 	mov    DWORD PTR ds:0x804a000,0x5
+```
+
+În fișierul executabil, codul instrucțiunilor conține adresa efectivă a variabilei `num_items` (`0x0804a000`), scrisă în format little endian.
+În fișierul obiect `one.o`, însă, nu apare adresa efectivă a variabilei `num_items`, ci apare `0x0`.
+Explicația este că fișierul executabil, rezultat în urma procesului de linking, a fost obținut știindu-se adresele efective, în vreme ce, în cazul fișierului obiect, adresele nu sunt cunoscute.
+
+Același lucru se întâmplă și în cazul instrucțiunii care referă simbolul `increment` din cadrul funcției `main`:
+```
+; one.o
+  [...]
+  20:	e8 fc ff ff ff       	call   21 <main+0xe>
+  [...]
+
+; one
+ [...]
+ 8048120:	e8 db ff ff ff       	call   8048100 <increment>
+ [...]
+```
+
+În fișierul executabil `one`, instrucțiunea `call` realizează saltul chiar la adresa (știută) a funcției `increment`.
+În fișierul obiect `one.o`, instrucțiunea `call` ar realiza un salt cumva chiar în interiorul său, la adresa `0x21`, instrucțiunea aflându-se ea însăși la adresa `0x20`.
+
+Modificarea instrucțiunilor sau a altor zone din viitorul executabil pentru a referi adresele efective ale simbolurilor folosite poartă numele de **relocare** (*relocation*).
+
+Din acest motiv, fișierele obiect mai sunt numite **fișiere obiect relocabile** (*relocatable object file*).
+Aceasta este explicația pentru care cuvântul *relocatable* apare în rezultatul rulării comenzilor `readelf` sau `file`:
+```
+[..]/01-one-file$ file one.o
+one.o: ELF 32-bit LSB relocatable, Intel 80386, version 1 (SYSV), with debug_info, not stripped
+
+[..]/01-one-file$ readelf -h one.o
+ELF Header:
+  [..]
+  Type:                              REL (Relocatable file)
+  [..]
+```
+
+Linkerul se ocupă de relocarea referințelor la simboluri.
+Atunci când creează fișierul executabil, urmărește referințele la simboluri și le înlocuiește cu adresele efective ale simbolurilor.
+Pentru a realiza relocarea, adică înlocuirea referințelor, linkerul trebuie să știe unde se găsesc aceste referințe.
+Pentru aceasta, fișierele obiect relocabile conțin **tabele de relocare** (*relocation tables*).
```

Comment:
Mie nu imi este foarte clar daca asta este definitia efectiva a unei tabele de relocare sau spre exemple got si plt pot fi vazute ca tabele de relocare. Sunt de acord ca pentru acest curs poate nu este foarte relevant, dar am vazut aceasta intrebare [1] in propunerea de teste de curs si m-a pus pe ganduri daca este nevoie de o definitie completa a ceea ce este un tabel de relocare.

[1] https://github.com/systems-cs-pub-ro/iocla-internal/pull/103/files#diff-b5a83fc3f4fac4c43634bbe86563cf42c1d3d58fd30c23deb48d6a25db37f56eR57

Reply:
Ambele sunt tabele de relocare.

Un executabil static nu are tabele de relocare. A folosit linkerul tabelele de relocare din fișierele obiect:
```
razvan@yggdrasil:~/.../iocla.git/curs/curs-18-19-linking$ readelf -r 01-one-file/one.o 

Relocation section '.rel.text' at offset 0x400 contains 4 entries:
 Offset     Info    Type            Sym.Value  Sym. Name
00000004  00000d01 R_386_32          00000000   num_items
0000000c  00000d01 R_386_32          00000000   num_items
00000018  00000d01 R_386_32          00000000   num_items
00000021  00000e02 R_386_PC32        00000000   increment
[...]

razvan@yggdrasil:~/.../iocla.git/curs/curs-18-19-linking$ readelf -r 01-one-file/one

There are no relocations in this file.
```

Un executabil static are tabelă de relocare pentru GOT/PLT (`.rel.plt`, `rel.got`). Aceste tabele de relocare sunt folosite de loader:
```
readelf -r 07-dynlib/main

Relocation section '.rel.dyn' at offset 0x37c contains 1 entry:
 Offset     Info    Type            Sym.Value  Sym. Name
08049ffc  00000306 R_386_GLOB_DAT    00000000   __gmon_start__

Relocation section '.rel.plt' at offset 0x384 contains 5 entries:
 Offset     Info    Type            Sym.Value  Sym. Name
0804a00c  00000107 R_386_JUMP_SLOT   00000000   increment
0804a010  00000207 R_386_JUMP_SLOT   00000000   print
0804a014  00000407 R_386_JUMP_SLOT   00000000   read
0804a018  00000507 R_386_JUMP_SLOT   00000000   init
0804a01c  00000607 R_386_JUMP_SLOT   00000000   __libc_start_main@GLIBC_2.0
```

Pot să adaug o definiție a [tabelei de relocare](https://en.wikipedia.org/wiki/Relocation_(computing)#Relocation_table), dar nu aș insista pe GOT / PLT, că alea sunt mai aproape de subiectul loading / dynamic linking pe care l-am trata la SO.